scipy/special/_evalpoly.h

```diff
@@ -0,0 +1,41 @@
+/* Evaluate polynomials.
+ *
+ * All of the coefficients are stored in reverse order, i.e. if the
+ * polynomial is
+ *
+ *     u_n x^n + u_{n - 1} x^{n - 1} + ... + u_0,
+ *
+ * then coeffs[0] = u_n, coeffs[1] = u_{n - 1}, ..., coeffs[n] = u_0.
+ *
+ * References
+ * ----------
+ * [1] Knuth, "The Art of Computer Programming, Volume II"
+ */
+
+#pragma once
+
+#include <complex>
+
+
+inline std::complex<double> cevalpoly(const double *coeffs, int degree,
+				      std::complex<double> z) {
+    /* Evaluate a polynomial with real coefficients at a complex point.
+     *
+     * Uses equation (3) in section 4.6.4 of [1]. Note that it is more
+     * efficient than Horner's method.
+     */
+    int j;
+    double a = coeffs[0];
+    double b = coeffs[1];
+    double r = 2 * z.real();
+    double s = std::norm(z);
+    double tmp;
+
+    for (j = 2; j < degree + 1; j++) {
```

Comment:
Rather then declare all variables at the top, I think it's best practice to declare them when you need them. And I don't believe there is a difference in speed, it just makes things much more readable.

So here you could do `for (int j = 2; ...` and `double tmp = b`.

Reply:
I'd argue that practice makes things really cumbersome. Having a single place to look is quite comforting when we are dealing with rather long code instead of hunting down where it was defined. 

I think this is one of the strengths of Fortran syntax which is very few. Also Cython has the same structure so we are quite familiar with it.

Reply:
I'm with @ilayn here. This is what we're more used to; I like having all of the variables declared at the top so you can see at a glance what variables will be used with their types. I understand that the standard practice in modern C++ is to declare variables right before they're used. I don't want to be like those people you hear about who "write Fortran in every language" though. Is this a really egregious violation of standard C++ style @izaid, or is it more just a matter of opinion?

Reply:
I wouldn't say it's egregious. But... :)

I would say what I described is the much more common style. Indeed, that's what's written in [ISOCpp](https://isocpp.org/wiki/faq/coding-standards) and [Google's C++ style guide](https://google.github.io/styleguide/cppguide.html). Readable is only one concern, there is an efficiency one as well. Empty declarations like those will also call the default constructor, though not for primitive types. Also certain types don't have a default constructor, and therefore can be used only when they are initialised. This includes references, e.g. `double &x` won't compile. Those are the reasons why it's more common to do it that way...

Anyway, that's my two cents!

Reply:
I changed it to declare the loop indices at first use, and fixed a place where I had an empty declaration and then put the definition on the next line. The variable `absz` could  be changed to be declared at first use in `lambertw` in `_lambertw.h`, but I don't think any of the others could be changed without torturing the code in some way, since each of `w`, `wz`, `wew` and `wewz` and `wn` are defined deep in nested logic, and in different ways in different branches.

Reply:
Again for `cevalpoly`, setting `double tmp = b` before the loop, would require moving `tmp = b` to the end of the loop, and if we didn't want to set `tmp` one more time than needed, we'd need to stop the loop one early, and then put 

`C++
	b = std::fma(-s, a, coeffs[j]);
	a = std::fma(r, a, tmp);
```

after the loop, which just doesn't seem worth it. I don't ike setting `tmp = b` an extra time either, even though it probably has no real performance impact.

Reply:
Well we are 1 function in and we already think the standard is rubbish so I guess nothing has changes since I stopped using C/C++ 😄 

Joking aside, I don't think this can actually matter since these are just a few bytes and not like large shaders or whatever. I'm pretty confident the standard is not talking to us with that paragraph.